src/fusion_trees/fusiontree.jl

```diff
@@ -0,0 +1,326 @@
+root_sector(f::SectorFusionTree) = f.root_sector
+branch_sectors(f::SectorFusionTree) = f.branch_sectors
+outer_multiplicity_indices(f::SectorFusionTree) = f.outer_multiplicity_indices
+
+# Base interface
+Base.convert(T::Type{<:Array}, f::SectorFusionTree) = convert(T, to_tensor(f))
+Base.isless(f1::SectorFusionTree, f2::SectorFusionTree) = isless(to_tuple(f1), to_tuple(f2))
+Base.length(::SectorFusionTree{<:Any,N}) where {N} = N
+
+# GradedUnitRanges interface
+GradedUnitRanges.sector_type(::SectorFusionTree{S}) where {S} = S
+
+function build_trees(legs::Vararg{AbstractGradedUnitRange})
+  tree_arrows = isdual.(legs)
+  sectors = blocklabels.(legs)
+  return mapreduce(vcat, CartesianIndices(blocklength.(legs))) do it
+    block_sectors = getindex.(sectors, Tuple(it))  # why not type stable?
+    return build_trees(block_sectors, tree_arrows)
+  end
+end
+
+# SymmetrySectors interface
+function SymmetrySectors.:×(f1::SectorFusionTree, f2::SectorFusionTree)
+  @assert arrows(f1) == arrows(f2)
+  product_leaves = .×(leaves(f1), leaves(f2))
+  product_root_sector = root_sector(f1) × root_sector(f2)
+  product_branch_sectors = .×(branch_sectors(f1), branch_sectors(f2))
+  product_outer_multiplicity_indices =
+    outer_multiplicity_kron.(
+      Base.tail(leaves(f1)),
+      branch_sectors(f1),
+      (Base.tail(branch_sectors(f1))..., root_sector(f1)),
+      outer_multiplicity_indices(f1),
+      outer_multiplicity_indices(f2),
+    )
+  return SectorFusionTree(
+    product_leaves,
+    arrows(f1),
+    product_root_sector,
+    product_branch_sectors,
+    product_outer_multiplicity_indices,
+  )
+end
+
+function SymmetrySectors.arguments(f::SectorFusionTree{<:SectorProduct})
+  transposed_indices =
+    outer_multiplicity_split.(
+      Base.tail(leaves(f)),
+      branch_sectors(f),
+      (Base.tail(branch_sectors(f))..., root_sector(f)),
+      outer_multiplicity_indices(f),
+    )
+  arguments_root = arguments(root_sector(f))
+  arguments_leaves = arguments.(leaves(f))
+  arguments_branch_sectors = arguments.(branch_sectors(f))
+  # TODO way to avoid explicit ntuple?
+  # works fine for Tuple and NamedTuple SectorProduct
+  return ntuple(
+    i -> SectorFusionTree(
+      getindex.(arguments_leaves, i),
+      arrows(f),
+      arguments_root[i],
+      getindex.(arguments_branch_sectors, i),
+      getindex.(transposed_indices, i),
+    ),
+    length(arguments_root),
+  )
+end
+
+function SymmetrySectors.arguments(f::SectorFusionTree{<:SectorProduct,0})
+  return map(arg -> SectorFusionTree((), (), arg, (), ()), arguments(root_sector(f)))
+end
+
+function SymmetrySectors.arguments(f::SectorFusionTree{<:SectorProduct,1})
+  arguments_root = arguments(root_sector(f))
+  arguments_leave = arguments(only(leaves(f)))
+  # use map(keys) to stay agnostic with respect to SectorProduct implementation
+  return map(
+    k -> SectorFusionTree((arguments_leave[k],), arrows(f), arguments_root[k], (), ()),
+    keys(arguments_root),
+  )
```

Comment:
```suggestion
  return map(keys(arguments_root)) do k
    return SectorFusionTree((arguments_leave[k],), arrows(f), arguments_root[k], (), ()),
  end
```

Reply:
You can also remove quite a few dict lookups by mapping over the `pairs` and not indexing `arguments_root`